pkg/pci/pci.go

```diff
@@ -155,27 +155,29 @@ func (i *Info) String() string {
 // New returns a pointer to an Info struct that contains information about the
 // PCI devices on the host system
 func New(opts ...*option.Option) (*Info, error) {
-	return NewWithContext(context.New(opts...))
-}
-
-// NewWithContext returns a pointer to an Info struct that contains information about
-// the PCI devices on the host system. Use this function when you want to consume
-// the topology package from another package (e.g. gpu)
-func NewWithContext(ctx *context.Context) (*Info, error) {
+	merged := option.Merge(opts...)
+	ctx := context.New(merged)
 	// by default we don't report NUMA information;
 	// we will only if are sure we are running on NUMA architecture
-	arch := topology.ARCHITECTURE_SMP
-	topo, err := topology.NewWithContext(ctx)
-	if err == nil {
-		arch = topo.Architecture
-	} else {
-		ctx.Warn("error detecting system topology: %v", err)
-	}
 	info := &Info{
-		arch: arch,
+		arch: topology.ARCHITECTURE_SMP,
 		ctx:  ctx,
 	}
-	if err := ctx.Do(info.load); err != nil {
+	// we do this trick because we need to make sure ctx.Setup() gets
+	// a chance to run before any subordinate package is created reusing
+	// our context.
+	if err := ctx.Do(func() error {
+		topo, err := topology.New(context.WithContext(ctx))
+		if err == nil {
+			info.arch = topo.Architecture
+		} else {
+			ctx.Warn("error detecting system topology: %v", err)
+		}
+		if context.Exists(merged) {
+			return info.load()
+		}
+		return ctx.Do(info.load)
+	}); err != nil {
```

Comment:
I'm deeply sorry but I managed to fumble this up. Now the order of operation is correct but we do things twice - the culprit is `ctx.Do` at the end. We need to reshuffle a bit the components: `ctx.Do`, anonymous func and `context.Exists`. This patch should fix:
```
diff --git a/pkg/pci/pci.go b/pkg/pci/pci.go
index 68d51b0..7411277 100644
--- a/pkg/pci/pci.go
+++ b/pkg/pci/pci.go
@@ -163,21 +163,27 @@ func New(opts ...*option.Option) (*Info, error) {
                arch: topology.ARCHITECTURE_SMP,
                ctx:  ctx,
        }
+
        // we do this trick because we need to make sure ctx.Setup() gets
        // a chance to run before any subordinate package is created reusing
        // our context.
-       if err := ctx.Do(func() error {
+       loadDetectingTopology := func() error {
                topo, err := topology.New(context.WithContext(ctx))
                if err == nil {
                        info.arch = topo.Architecture
                } else {
                        ctx.Warn("error detecting system topology: %v", err)
                }
-               if context.Exists(merged) {
-                       return info.load()
-               }
-               return ctx.Do(info.load)
-       }); err != nil {
+               return info.load()
+       }
+
+       var err error
+       if context.Exists(merged) {
+               err = loadDetectingTopology()
+       } else {
+               err = ctx.Do(loadDetectingTopology)
+       }
+       if err != nil {
                return nil, err
        }
        return info, nil
```
IOW the function should read like
```
func New(opts ...*option.Option) (*Info, error) {
	merged := option.Merge(opts...)
	ctx := context.New(merged)
	// by default we don't report NUMA information;
	// we will only if are sure we are running on NUMA architecture
	info := &Info{
		arch: topology.ARCHITECTURE_SMP,
		ctx:  ctx,
	}

	// we do this trick because we need to make sure ctx.Setup() gets
	// a chance to run before any subordinate package is created reusing
	// our context.
	loadDetectingTopology := func() error {
		topo, err := topology.New(context.WithContext(ctx))
		if err == nil {
			info.arch = topo.Architecture
		} else {
			ctx.Warn("error detecting system topology: %v", err)
		}
		return info.load()
	}

	var err error
	if context.Exists(merged) {
		err = loadDetectingTopology()
	} else {
		err = ctx.Do(loadDetectingTopology)
	}
	if err != nil {
		return nil, err
	}
	return info, nil
}
```

And I verified carefully this version and looks really good this time, so third time should be the charm!

Reply:
And I'm thinking about a test to catch this. I'll add in a followup PR.